Add On time blog (#181)
openfaas/openfaas.github.io: _posts/2020-04-20-schedule-your-functions.md

---

> title: "On time - schedule your OpenFaaS functions"

Reviewer: I like the "on time" pun, what if you put it into the intro rather than the description?

Something like this might be more descriptive and better for SEO:

"Keep your functions running to schedule with the cron-connector."

Author: Yeah ok :+1: Done :heavy_check_mark:

> description: "In this post Martin will show you how to deploy a component which will execute your functions on a schedule"

Reviewer: "In this post Martin will show you how schedule your functions with the cron-connector, for regular tasks and scheduled jobs"

Reviewer: (Adding the why)

Author: Sounds better, changed :heavy_check_mark:

date: 2020-04-20
image: /images/2020-04-20-schedule-your-functions/accuracy-alarm-clock-analogue.jpg
categories:

>   - cronjobs

Reviewer: Just `cron`?

Author: Yep, changed :+1:

  - kubernetes
  - serverless
  - linux
  - tutorial
  - example
author_staff_member: martin
dark_background: true

---

> In this blog post I will show you how you can run your functions on schedule. We will first set our environment and then we will run the cron-connector with example function which will be executed on a schedule.

Reviewer: This line should be the same as "description" IMHO. Don't mind which is which, that's up to you, but make them the same.

Author: The description replaces this line now :+1:

One of the things we need very often is our functions to be executed on a schedule. This is where the cron-connector shines. We will showcase how you can use it in this post.

## Motivation

OpenFaaS provides a very convenient and flexible way to quickly write and expose your code to be used by external systems. Often times those systems don't have a hook to which you can schedule function execution. Also often times we need to run those functions without external systems, we just want our code to be ran on an interval.

For those purposes [connector-sdk](https://github.com/openfaas-incubator/connector-sdk) was created. It provides the means to quickly adapt different systems to invoke functions. Some of them are [Kafka](https://github.com/openfaas-incubator/kafka-connector), [NATS](https://github.com/openfaas-incubator/nats-connector), [AWS SNS](https://github.com/affix/OpenFaaS-SNS) and [many more](https://docs.openfaas.com/reference/triggers/).

> Now we need our functions to to be invoked on an interval. In Linux we have the [cron configuration](https://en.wikipedia.org/wiki/Cron) that allows us to execute code on a schedule. In Kubernetes, this was abstracted into the `CronJob` object.  For any OpenFaaS cluster, the [cron-connector](https://github.com/openfaas-incubator/cron-connector) uses the `connector-sdk` to provide the same functionality, running our functions on a predefined schedule.

Reviewer: cron configuration -> cron

Author: Swapped :heavy_check_mark:

```

Wait for the installation to finish and verify the pods in the `kube-system` namespace are `1/1`:

```bash
kubectl get pods -n kube-system
```

### Install OpenFaaS

With `arkade` the installation of OpenFaaS boils down to single line command:

```bash
arkade install openfaas
```

Wait for the OpenFaaS gateway to be ready:

```bash
kubectl rollout status -n openfaas deploy/gateway
```

### Set up OpenFaaS client

Get the OpenFaaS Gateway password:

```bash
PASSWORD=$(kubectl get secret -n openfaas basic-auth -o jsonpath="{.data.basic-auth-password}" | base64 --decode; echo)
```

Set up the gateway endpoint for the client:

```bash
export OPENFAAS_URL=http://$(docker inspect -f "{{ .NetworkSettings.IPAddress }}" $(docker ps -aqf "name=kind-control-plane")):31112
```

Authenticate the client:

```bash
echo -n $PASSWORD | faas-cli login --username admin --password-stdin
```

## Schedule the functions

After we have the environment up and running, now we can install the cron-connector and schedule our functions to run at specific intervals.

### Deploy the cron-connector

In order to install our `cron-connector` and schedule a function run the following command:

> ```bash

Reviewer: Use arkade to install the connector, there is an app for it.

```
Available Commands:
  cert-manager            Install cert-manager
  chart                   Install the specified helm chart
  cron-connector          Install cron-connector for OpenFaaS
```

Author: Installed with arkade now :+1:

curl -s https://raw.githubusercontent.com/openfaas-incubator/cron-connector/master/yaml/kubernetes/connector-dep.yml | kubectl apply -f -
```

Wait for the `cron-connector` to be deployed:

```bash
kubectl rollout status -n openfaas deploy/sample-connector
```

## Run the function

> We will run a function from the store once a minute with a `schedule` annotation following this cron syntax `*/1 * * * *` which is the same as in the [linux cron configuration](https://www.cyberciti.biz/faq/how-do-i-add-jobs-to-cron-under-linux-or-unix-oses/).

Reviewer: "which is the same as in the [linux cron configuration]" -> See this blog post for some examples of the cron format.

Author: Changed as suggested :+1:

> The cron-connector recognizes which functions should be ran on a schedule by annotation `topic` with value `cron-function`.
> 
> Finally we will deploy `node-info` function and apply the schedule and topic annotations mentioned above:

Reviewer 2: I would merge these two paragraphs

```suggestion
Finally we will deploy `node-info` function . The cron-connector recognizes which functions should be ran on a schedule by annotation `topic` with value `cron-function`, we can add this annotation to the function during deployment like this:
```

Author: Changed that Lucas, thanks :+1:

> ```bash

Reviewer: Please use a function that does something practical.

The uptime monitor idea you had seems like a good idea. Or scraping a webpage to see if some data is present.

scrappy is a good HTML scraping module for python for instance.

Author: I will develop the function and add it in place of this one :+1:

faas-cli store deploy nodeinfo --annotation schedule="*/1 * * * *" --annotation topic=cron-function
```

If you wait for a few minutes and visit the OpenFaaS UI you will see how the `Invocation count` of the function raises every minute.

## Wrapping up

In this post we have shown how you can run minimal OpenFaaS environment with simple function and how with the help of the cron-connector you can execute the function on a desired schedule.

> Now you can go ahead and [create your own function](https://docs.openfaas.com/cli/templates/) in the limitless amount of languages and make sure it always runs on time!

Reviewer: "in the limitless amount of languages and make sure it always runs on time!" -> Can you let the reader decide if it's limitless?

How about:

```
Learn how to create your own functions with the [OpenFaaS workshop]. You'll find many templates in the template store, run `faas-cli template store list` or view the [Template Documentation]
```

Author: Changed as suggested :+1:
